### PR TITLE
Feat hausdorff distance

### PR DESCRIPTION
**Pull request by benoit2776**

### CODE REVIEW

File: include/geode/mesh/helpers/hausdorff_distance.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#pragma once
+
+#include <geode/mesh/common.h>
+
+namespace geode
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( AABBTree );
+    FORWARD_DECLARATION_DIMENSION_CLASS( TriangulatedSurface );
+    FORWARD_DECLARATION_DIMENSION_CLASS( DistanceToTriangle );
```
Comment (BotellaA) [inline include/geode/mesh/helpers/hausdorff_distance.h:32]:
```suggestion
    ALIAS_3D( TriangulatedSurface );
```

File: tests/mesh/test-hausdorff-distance.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include <geode/mesh/core/triangulated_surface.h>
+#include <geode/mesh/helpers/aabb_surface_helpers.h>
+#include <geode/mesh/helpers/hausdorff_distance.h>
+#include <geode/mesh/io/triangulated_surface_input.h>
+
+#include <geode/tests/common.h>
+
+void test()
+{
+    const std::string filename( "../tests/data/iter1.og_tsf3d" );
```
Comment (BotellaA) [inline tests/mesh/test-hausdorff-distance.cpp:16]: use this https://github.com/Geode-solutions/OpenGeode/blob/master/tests/mesh/test-graph.cpp#L168

File: tests/mesh/test-hausdorff-distance.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_A =
+        geode::load_triangulated_surface< 3 >( filename );
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_B =
+        geode::load_triangulated_surface< 3 >(
+            "../tests/data/modified_iter1.og_tsf3d" );
+
+    const auto hausdorff_distance =
+        HAUSDORFF_DISTANCE::compute_two_sided_hausdorff_distance(
+            *mesh_A, *mesh_B );
+    DEBUG( hausdorff_distance );
```
Comment (BotellaA) [inline tests/mesh/test-hausdorff-distance.cpp:26]: use OPENGEODE_EXCEPTION to check the value to its answer

File: include/geode/mesh/helpers/hausdorff_distance.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+} // namespace geode
+
+namespace HAUSDORFF_DISTANCE
+{
+    double compute_one_sided_hausdorff_distance(
```
Comment (BotellaA) [inline include/geode/mesh/helpers/hausdorff_distance.h:37]: @panquez do we want both these functions, or a single one `hausdorff_distance( mesh, mesh )`?

Comment (panquez) [reply include/geode/mesh/helpers/hausdorff_distance.h:37]: yes I think only the two sided H distance is necessary

File: include/geode/mesh/helpers/hausdorff_distance.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#include <geode/mesh/common.h>
+
+namespace geode
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( AABBTree );
+    FORWARD_DECLARATION_DIMENSION_CLASS( TriangulatedSurface );
+    FORWARD_DECLARATION_DIMENSION_CLASS( DistanceToTriangle );
+} // namespace geode
+
+namespace HAUSDORFF_DISTANCE
```
Comment (panquez) [inline include/geode/mesh/helpers/hausdorff_distance.h:35]: We do not use custom namespace in OpenGeode. We only use : anonymous (but only in cpp) or geode namespace

File: include/geode/mesh/helpers/hausdorff_distance.h (hunk 4 above)
Comment (panquez) [reply include/geode/mesh/helpers/hausdorff_distance.h:37]: you can remove "compute" in function name

File: tests/mesh/test-hausdorff-distance.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+#include <geode/tests/common.h>
+
+void test()
+{
+    const std::string filename( "../tests/data/iter1.og_tsf3d" );
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_A =
+        geode::load_triangulated_surface< 3 >( filename );
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_B =
+        geode::load_triangulated_surface< 3 >(
+            "../tests/data/modified_iter1.og_tsf3d" );
```
Comment (panquez) [inline tests/mesh/test-hausdorff-distance.cpp:21]: Do we use iter1 as data or more neutral data like Armadillo or another one?

File: tests/mesh/test-hausdorff-distance.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+#include <geode/mesh/io/triangulated_surface_input.h>
+
+#include <geode/tests/common.h>
+
+void test()
+{
+    const std::string filename( "../tests/data/iter1.og_tsf3d" );
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_A =
+        geode::load_triangulated_surface< 3 >( filename );
+    std::unique_ptr< geode::TriangulatedSurface3D > mesh_B =
```
Comment (panquez) [inline tests/mesh/test-hausdorff-distance.cpp:19]: const auto

File: src/geode/mesh/helpers/hausdorff_distance.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace HAUSDORFF_DISTANCE
+{
+
+    double compute_one_sided_hausdorff_distance(
+        const geode::TriangulatedSurface< 3 >& mesh_A,
+        const geode::TriangulatedSurface< 3 >& mesh_B )
+    {
+        const auto mesh_B_tree = geode::create_aabb_tree( mesh_B );
+        const geode::DistanceToTriangle< 3 > distance_action( mesh_B );
```
Comment (panquez) [inline src/geode/mesh/helpers/hausdorff_distance.cpp:23]:
```suggestion
        const geode::DistanceToTriangle3D distance_action{ mesh_B };
```

File: src/geode/mesh/helpers/hausdorff_distance.cpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        double min_dist;
+        for( const geode::index_t v : geode::Range( mesh_A.nb_vertices() ) )
```
Comment (panquez) [inline src/geode/mesh/helpers/hausdorff_distance.cpp:25]:
```suggestion
        for( auto v : geode::Range{ mesh_A.nb_vertices() } )
```

File: src/geode/mesh/helpers/hausdorff_distance.cpp [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            const geode::Point3D& query = mesh_A.point( v );
```
Comment (panquez) [inline src/geode/mesh/helpers/hausdorff_distance.cpp:27]:
```suggestion
            const auto& query = mesh_A.point( v );
```

File: src/geode/mesh/helpers/hausdorff_distance.cpp [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+            const auto toto =
+                mesh_B_tree.closest_element_box( query, distance_action );
+            const auto distance = std::get< 2 >( toto );
```
Comment (panquez) [inline src/geode/mesh/helpers/hausdorff_distance.cpp:30]: toto ?

**Review by panquez — CHANGES_REQUESTED** [review]

**Review by BotellaA — APPROVED** [review]

File: tests/mesh/test-hausdorff-distance.cpp (hunk 3 above)
Comment (Ply1501) [reply tests/mesh/test-hausdorff-distance.cpp:27]: [](https://www.dropbox.com/s/epjibi1epidw9bm/Fichier%2009-07-2022%2002%2001%2049.txt?dl=0)

**Pull request merged**